lib/Sema/DerivedConformanceEquatableHashable.cpp

```diff
@@ -1336,28 +1336,50 @@ ValueDecl *DerivedConformance::deriveHashable(ValueDecl *requirement) {
     if (hashValueDecl->isImplicit()) {
       // Neither hashValue nor hash(into:) is explicitly defined; we need to do
       // a full Hashable derivation.
-      
+
       // Refuse to synthesize Hashable if type isn't a struct or enum, or if it
       // has non-Hashable stored properties/associated values.
       auto hashableProto = C.getProtocol(KnownProtocolKind::Hashable);
+      auto *conformanceDC = ConformanceDecl->getDeclContext();
+      auto nominalTy = Nominal->getDeclaredType();
       if (!canDeriveConformance(getConformanceContext(), Nominal,
                                 hashableProto)) {
-        ConformanceDecl->diagnose(diag::type_does_not_conform,
-                                  Nominal->getDeclaredType(),
+        ConformanceDecl->diagnose(diag::type_does_not_conform, nominalTy,
                                   hashableProto->getDeclaredType());
         // Ideally, this would be diagnosed in
         // ConformanceChecker::resolveWitnessViaLookup. That doesn't work for
         // Hashable because DerivedConformance::canDeriveHashable returns true
         // even if the conformance can't be derived. See the note there for
         // details.
-        auto *dc = ConformanceDecl->getDeclContext();
-        tryDiagnoseFailedHashableDerivation(dc, Nominal);
+        tryDiagnoseFailedHashableDerivation(conformanceDC, Nominal);
         return nullptr;
       }
 
       if (checkAndDiagnoseDisallowedContext(requirement))
         return nullptr;
 
+      // Warn the user that having a custom == almost surely also require a
+      // custom Hashable conformance even if for source stability reasons
+      // it will be synthesized
+      if (auto eqConformance = TypeChecker::conformsToProtocol(
+              conformanceDC->mapTypeIntoContext(nominalTy),
+              C.getProtocol(KnownProtocolKind::Equatable), conformanceDC,
+              None)) {
+        DeclName equalsName(C, DeclBaseName(C.Id_EqualsOperator),
+                            { Identifier(), Identifier() });
+        ConcreteDeclRef witness =
+            eqConformance->getWitnessByName(nominalTy->getRValueType(),
+                                            equalsName);
+
+        auto equalsDeclaration = witness.getDecl();
+        if (!equalsDeclaration->isImplicit()) {
```

Comment:
```suggestion
        if (equalsDeclaration && !equalsDeclaration->isImplicit()) {
```

just in case we don't have a decl. Also, check if it fixes that crash

Reply:
I added the check but the crash remains, but I’ve noted that I’ve missed a line in the stack trace `Assertion failed: (!witnessDC->isInnermostContextGeneric()), function getWitnessDeclRef, file /Users/jacopo/Codes/swift-source/swift/lib/AST/ProtocolConformance.cpp, line 947.
Stack dump:`
Seems that I’m using something wrong in getting the witness of the `==` declaration. But I don’t get what is wrong.

Reply:
Can you run the compiler with the debugger attached (copy the giant command-line dump you saw in the backtrace, then put `lldb` in front of it and `--` after the executable name, then execute it and type "run") and then, once it crashes, use the command `e witnessDC->dumpContext()` to see what the witnessDC is? That might give us a better idea of what's going wrong.

Reply:
The dump is this:
```
0x118804ea0 Module name=Swift
  0x118804f60 FileUnit file="/swift-source/build/Ninja-RelWithDebInfoAssert+swift-DebugAssert/swift-macosx-x86_64/lib/swift/macosx/Swift.swiftmodule/x86_64.swiftmodule"
    0x1188f0948 StructDecl name=Int
      0x1188e3a90 AbstractFunctionDecl name=init(_builtinIntegerLiteral:) : (Int.Type) -> (IntLiteral) -> Int
```

Reply:
In that point it seems that witnessDC is:
```
(lldb) e witnessDC->dumpContext()
0x118815ca0 Module name=Swift
  0x118815d60 FileUnit file="/swift-source/build/Ninja-RelWithDebInfoAssert+swift-DebugAssert/swift-macosx-x86_64/lib/swift/macosx/Swift.swiftmodule/x86_64.swiftmodule"
    0x118044b40 AbstractFunctionDecl name===(_:_:) : <T where T : RawRepresentable, T.RawValue : Equatable> (T, T) -> Bool
```

I've tried to hunt up the stack to gather some other context, these are the variables inside my block in DerivedConformanceEquatableHashable.cpp
```
e eqConformance->dump()
(normal_conformance type=TestRawType protocol=Equatable
  (value req===(_:_:) witness=Swift.(file).==))
e nominalTy->dump()
(enum_type decl=main.(file).TestRawType@/swift-source/swift/test/Sema/implementation-only-import-in-decls.swift:65:13)
```

After stepping in in the getWitnessByName function I found these:
```
e type.dump()
(enum_type decl=main.(file).TestRawType@/swift-source/swift/test/Sema/implementation-only-import-in-decls.swift:65:13)

e name.getBaseName()
(swift::DeclBaseName) $0 = {
  Ident = (Pointer = "==")
}

(lldb) e getConcrete()->dump()
(normal_conformance type=TestRawType protocol=Equatable
  (value req===(_:_:) witness=Swift.(file).==))

e witness
(swift::Witness) $1 = {
  storage = {
    llvm::pointer_union_detail::PointerUnionMembers<llvm::PointerUnion<swift::ValueDecl *, swift::Witness::StoredWitness *>, llvm::PointerIntPair<void *, 1, int, llvm::pointer_union_detail::PointerUnionUIntTraits<swift::ValueDecl *, swift::Witness::StoredWitness *>, llvm::PointerIntPairInfo<void *, 1, llvm::pointer_union_detail::PointerUnionUIntTraits<swift::ValueDecl *, swift::Witness::StoredWitness *> > >, 0, swift::ValueDecl *, swift::Witness::StoredWitness *> = {
      llvm::pointer_union_detail::PointerUnionMembers<llvm::PointerUnion<swift::ValueDecl *, swift::Witness::StoredWitness *>, llvm::PointerIntPair<void *, 1, int, llvm::pointer_union_detail::PointerUnionUIntTraits<swift::ValueDecl *, swift::Witness::StoredWitness *>, llvm::PointerIntPairInfo<void *, 1, llvm::pointer_union_detail::PointerUnionUIntTraits<swift::ValueDecl *, swift::Witness::StoredWitness *> > >, 1, swift::Witness::StoredWitness *> = {
        llvm::pointer_union_detail::PointerUnionMembers<llvm::PointerUnion<swift::ValueDecl *, swift::Witness::StoredWitness *>, llvm::PointerIntPair<void *, 1, int, llvm::pointer_union_detail::PointerUnionUIntTraits<swift::ValueDecl *, swift::Witness::StoredWitness *>, llvm::PointerIntPairInfo<void *, 1, llvm::pointer_union_detail::PointerUnionUIntTraits<swift::ValueDecl *, swift::Witness::StoredWitness *> > >, 2> = {
          Val = (Value = 4700503004)
        }
      }
    }
  }
}
```

Reply:
I’ve recreated the problem, it seems that if the `Equatable` conformance is declared in a different module, and the type extend that, the `eqConformance` is not valid to generate a witness.
Anyone knows how can I catch that? Using lldb I haven’t find much differences between the tests that I can use to disambiguate them and the methods of `eqConformance` don’t seems to have that particular catch available, or at least not to my eyes.

I think that the correct flow is that in the if that checks that the conformance exists must check even if the conformance is inside my module before finding the witness (even because is nonsensical to add a warning in another module right?).